### PR TITLE
DAT-336-fix-RemoveNoneOrEmptyElement-to-remove-None-element-completely

### PR DESCRIPTION
**Pull request by tunglxfast**
Current converter RemoveNoneOrEmptyElement will remove None key/value to empty dict, which should be remove completely.
[DAT-336](https://rapiddweller.atlassian.net/browse/DAT-336)

[DAT-336]: https://rapiddweller.atlassian.net/browse/DAT-336?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: datamimic_ce/converter/remove_none_or_empty_element_converter.py
Comment (ake2l) [inline datamimic_ce/converter/remove_none_or_empty_element_converter.py:1]:
```py
import copy
from typing import Dict, List, Union

from datamimic_ce.converter.converter import Converter

EMPTY_VALUES = {None, {}, [], ""}

def is_empty(value) -> bool:
    return value in EMPTY_VALUES

class RemoveNoneOrEmptyElementConverter(Converter):
    """
    Remove None or empty elements from a list or dictionary, including nested structures.
    """

    def convert(self, value: Union[List, Dict]) -> Union[List, Dict]:
        return self._clean(value)

    def _clean(self, value: Union[List, Dict]) -> Union[List, Dict]:
        if isinstance(value, list):
            return self._clean_list(value)
        elif isinstance(value, dict):
            return self._clean_dict(value)
        return value

    def _clean_list(self, value: List) -> List:
        return [self._clean(item) for item in value if not is_empty(item)]

    def _clean_dict(self, value: Dict) -> Dict:
        return {key: self._clean(val) for key, val in value.items() if not is_empty(val)}
```

can you try to optimize the code a bit ?

or something like this ? 

```py
from typing import Dict, List, Union, Any, Set
import copy

class RemoveNoneOrEmptyElementConverter(Converter):
    """Removes None or empty elements from nested data structures efficiently."""
    
    __slots__ = ()  # Optimize memory usage
    
    # Use frozenset for faster lookups
    EMPTY_VALUES: frozenset = frozenset((None, "", [], {}))
    
    def convert(self, value: Union[List[Any], Dict[str, Any]]) -> Union[List[Any], Dict[str, Any]]:
        """Recursively removes empty elements."""
        # Early return for primitive types
        if not isinstance(value, (list, dict)):
            return value
            
        # Dispatch to specialized handlers
        return (self._convert_list(value) if isinstance(value, list) 
                else self._convert_dict(value))

    def _convert_list(self, values: List[Any]) -> List[Any]:
        """Process list elements with list comprehension."""
        # Single-pass list comprehension with inline conversion
        return [
            converted for item in values
            if item not in self.EMPTY_VALUES
            and (converted := self.convert(item)) not in self.EMPTY_VALUES
        ]

    def _convert_dict(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Process dictionary with minimal copying."""
        # Create new dict directly with filtered items
        result = {}
        for key, value in data.items():
            if value not in self.EMPTY_VALUES:
                converted = self.convert(value)
                if converted not in self.EMPTY_VALUES:
                    result[key] = converted
        return result
```

**Review by ake2l — CHANGES_REQUESTED** [review]
and please try to add also isolated unittest, i would like to extend the amount of unittest we have ...

```py
import pytest
from typing import Any, Dict, List

@pytest.fixture
def converter():
    return RemoveNoneOrEmptyElementConverter()

class TestRemoveNoneOrEmptyElementConverter:
    @pytest.mark.parametrize("input_val,expected", [
        (42, 42),
        (3.14, 3.14),
        ("test", "test"),
        ("", ""),
        (True, True),
        (None, None),
    ])
    def test_primitive_types(self, converter, input_val: Any, expected: Any):
        assert converter.convert(input_val) == expected

    @pytest.mark.parametrize("input_list,expected", [
        ([1, None, [], "", {}, "test"], [1, "test"]),
        ([], []),
        ([None, [], {}, ""], []),
        ([[[[]]]], []),
    ])
    def test_list_handling(self, converter, input_list: List, expected: List):
        assert converter.convert(input_list) == expected

    @pytest.mark.parametrize("input_dict,expected", [
        ({"a": 1, "b": None, "c": [], "d": ""}, {"a": 1}),
        ({}, {}),
        ({"a": None, "b": [], "c": {}}, {}),
        ({"a": {"b": {"c": []}}}, {}),
    ])
    def test_dict_handling(self, converter, input_dict: Dict, expected: Dict):
        assert converter.convert(input_dict) == expected

    def test_nested_structures(self, converter):
        input_data = {
            "list": [1, None, {"a": None, "b": 2}],
            "dict": {"x": [], "y": [1, None, ""]},
            "empty": {}
        }
        expected = {
            "list": [1, {"b": 2}],
            "dict": {"y": [1]}
        }
        assert converter.convert(input_data) == expected

    def test_complex_mixed_types(self, converter):
        input_data = [
            1,
            ["nested", None, []],
            {"a": 1, "b": None, "c": {"d": [], "e": 2}},
            None,
            [],
            [[], {}, None]
        ]
        expected = [
            1,
            ["nested"],
            {"a": 1, "c": {"e": 2}}
        ]
        assert converter.convert(input_data) == expected

    @pytest.mark.slow
    def test_large_structure_performance(self, converter):
        large_input = {
            str(i): [None, [], {}] if i % 2 else [1, {"x": i}]
            for i in range(1000)
        }
        result = converter.convert(large_input)
        assert all(isinstance(v, list) and len(v) > 0 for v in result.values())

    def test_circular_reference(self, converter):
        circular_list = [1, 2]
        circular_list.append(circular_list)
        
        with pytest.raises(RecursionError):
            converter.convert(circular_list)

    @pytest.mark.parametrize("input_val", [
        [1, [2, [3, [4, [5]]]]],  # Deep nesting
        {"a": {"b": {"c": {"d": {"e": 1}}}}},  # Deep dict
        [{"a": [{"b": []}]}, [{"c": {}}]],  # Mixed deep structure
    ])
    def test_deep_structures(self, converter, input_val):
        result = converter.convert(input_val)
        assert result  # Ensure non-empty result

```

**Comment by tunglxfast** [comment]
@ake2l Hi Alex, I changed RemoveNoneOrEmptyElementConverter class to 2nd option that you gave, and change the test. Please check it again, thank you.

**Review by ake2l — APPROVED** [review]
well done mr tung 👍🏽 , next time please use pytest for the unittest

**Pull request merged**